networking/src/Ntp/Client.hs

```diff
@@ -77,280 +86,239 @@ data NtpClient = NtpClient
-
-logDebug :: Text -> IO ()
-logDebug = usingNtpLogger . Wlog.logDebug
+-- |
+-- Update @'ncStatus'@ according to received responses.
+updateStatus'
+    :: NtpClient
+    -> ([NtpOffset] -> (NtpStatus, (Wlog.Severity, Text)))
+    -> IO ()
+updateStatus' cli fn = do
+    (offset, msg) <- fn <$> readTVarIO (ncState cli)
+    traceWith ntpTrace msg
+    atomically $ writeTVar (ncStatus cli) offset
+
+updateStatus :: NtpClient -> IO ()
+updateStatus cli = updateStatus' cli fn
+    where
+    fn :: [NtpOffset]
+       -> (NtpStatus, (Wlog.Severity, Text))
+    fn [] = ( NtpSyncUnavailable
+            , (Wlog.Warning, "ntp client haven't received any response")
+            )
+    fn offsets =
+        let offset = ntpSelection (ncSettings cli) $ NE.fromList $ offsets
+        in ( NtpDrift offset
+           , (Wlog.Info, sformat ("Evaluated clock offset "%shown%"mcs") offset)
+           )
 
 -- |
--- Handle results.  It is either when all ntp servers responded or when
--- `ntpResponseTimeout` has passed since the request where send.  If none of the servers responded `ncState`
-handleCollectedResponses :: NtpClient -> IO ()
-handleCollectedResponses cli = do
-    mres <- readTVarIO (ncState cli)
-    case mres of
-        Nothing        -> do
-            atomically $ writeTVar (ncStatus cli) NtpSyncUnavailable
-            logError "Protocol error: responses are not awaited"
-        Just []        -> do
-            atomically $ writeTVar (ncStatus cli) NtpSyncUnavailable
-            logWarning "No server responded"
-        Just responses -> handleE `handleAny` do
-            let ntpOffset = ntpSelection (ncSettings cli) $ NE.fromList $ responses
-            logInfo $ sformat ("Evaluated clock offset "%shown%"mcs") (toMicroseconds ntpOffset)
-            atomically $ writeTVar (ncStatus cli) (NtpDrift $ ntpOffset)
-  where
-    handleE = logError . sformat ("ntpSelection: "%shown)
+-- Every `ntpPollDelay` we send a request to the list of `ntpServers`.  Before
+-- sending a request, we put `NtpSyncPending` to `ncState`.  After sending
+-- all requests we wait until either all servers responded or
+-- `ntpResponseTimeout` passesed.  If at least one server responded
+-- `handleCollectedResponses` will update `ncStatus` in `NtpClient` with a new
+-- drift.
+sendLoop :: NtpClient -> [Addresses] -> IO ()
+sendLoop cli addrs = do
 
 
-allResponsesGathered :: NtpClient -> STM Bool
-allResponsesGathered cli = do
-    responsesState <- readTVar $ ncState cli
-    let servers = ntpServers $ ncSettings cli
-    return $ case responsesState of
-        Nothing        -> False
-        Just responses -> length responses >= length servers
+    let respTimeout = ntpResponseTimeout (ncSettings cli)
+    let poll        = ntpPollDelay (ncSettings cli)
+
+    () <- withAsync
+        (do
+            -- wait for responses and update status
+            _ <- timeout respTimeout waitForResponses
+            updateStatus cli
```

Comment:
Good point, I was actually thinking about doing this too.  I haven't done it because we need `ncState :: NtpClient -> TVar [NtpOffset]` variable to communicate between the listener thread and sending thread.  But we can change how the client is run: if we start the listener thread in this place we will not need this tvar.  The communication will be done just using `ncStatus`.

Reply:
We need to listen on two sockets, so I think there's no way around impure `ncState`. Maybe you have some other idea?

Reply:
I wonder if having a receiving thread (`startReceive`) is the wrong way around. Do we ever expect to get a message without first sending one? That's to say, are we acting as an ntp *server* as well as a client? If not, maybe we should `recvFrom` the socket after sending on it. That would give a simpler implementation, with less concurrency and fewer mutable cells.

Reply:
It is, but I haven't changed that.  I just didn't want to rewrite everything. I will add a todo, and let's move on.